File: 2516. Take K of Each Character From Left and Right.java
// Medium
// Sliding Window, Counting
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/take-k-of-each-character-from-left-and-right/

class Solution {
    public int takeCharacters(String s, int k) {
        int n = s.length(), res = n;
        int[] count = new int[26];
        for (int i = 0; i < n; i++) {
            count[s.charAt(i) - 'a']++;
        }
        if (count[0] >= k && count[1] >= k && count[2] >= k) {
            res = Math.min(res, n);
        } else {
            return -1;
        }
        int left = 0;
        for (int right = 0; right < n; right++) {
            count[s.charAt(right) - 'a']--;
            while (left < right && (count[0] < k || count[1] < k || count[2] < k)) {
                count[s.charAt(left) - 'a']++;
                left++;
            }
            if (count[0] >= k && count[1] >= k && count[2] >= k) {
                res = Math.min(res, n - (right - left + 1));
            }
        }
        return res;
    }
}

File: 2001-3000/2516. Take K of Each Character From Left and Right.py
# Time:O(n), Space:O(1)

from collections import Counter

class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        c_count = Counter(s)
        if c_count["a"] < k or c_count["b"] < k or c_count["c"] < k:
            return -1

        n = len(s)
        max_window = 0
        left = 0
        window_count = Counter()

        for right in range(n):
            window_count[s[right]] += 1

            while (
                c_count["a"] - window_count["a"] < k
                or c_count["b"] - window_count["b"] < k
                or c_count["c"] - window_count["c"] < k
            ):
                window_count[s[left]] -= 1
                left += 1

            max_window = max(max_window, right - left + 1)

        return n - max_window
    